### PR TITLE
Plando: Fix placing various items on invalid locations

### PR DESCRIPTION
**Pull request by flagrama**
When using the invert pattern in plandomizer (`!Bow`) on a non-song location and with songsanity off ignore the song item pool when choosing the item to place. It should only do this if a pattern is used to avoid placing an item the user did not intend.

Prior to this change trying to generate 100 seeds with an invert pattern would cause an error due to trying to place a song on a chest location part way through. After this change I could generate over 200 without issues.

The same is done for songs as a user using an invert pattern with songsanity off likely doesn't want to place an item on the song location, and would cause an error anyway since the pool is not modified to remove a song.

Not sure if it is needed for dungeon items being placed with any settings as I haven't run into it myself.

Would probably be a good idea to set up some kind of enum or something for pools so I'm not using 2 (Song pool) as a magic number in case pools ever change.

### CODE REVIEW

**Comment by flagrama** [comment]
The newest commit does stop non-dungeon rewards on dungeon reward locations and vice versa, however it also stops raising an error on unknown locations due to typos and such currently.

**Comment by flagrama** [comment]
And that one should fix that.

File: Plandomizer.py [hunk 1]
```diff
@@ -532,26 +534,38 @@ def fill(self, window, worlds, location_pools, item_pools):
...
+
             if record.item == '#Junk' and location.type == 'Song' and not world.shuffle_song_items:
                 record.item = '#JunkSong'
 
+            ignore_pools = None
+            is_invert = pattern_matcher(record.item)('!')
+            if is_invert and location.type != 'Song' and not world.shuffle_song_items:
+                ignore_pools = [2]
+            if is_invert and location.type == 'Song' and not world.shuffle_song_items:
+                ignore_pools = [i for i, pools in enumerate(item_pools) if i != 2]
```
Comment (Zannick) [inline Plandomizer.py]: range(len(...)) instead of enumerate?

File: Plandomizer.py [hunk 2]
```diff
@@ -532,26 +534,38 @@ def fill(self, window, worlds, location_pools, item_pools):
...
+            if record.item in item_groups['DungeonReward']:
+                raise RuntimeError('Cannot place dungeon reward %s in world %d in location %s.' % (record.item, self.id + 1, location_name))
+
             if record.item == '#Junk' and location.type == 'Song' and not world.shuffle_song_items:
                 record.item = '#JunkSong'
 
+            ignore_pools = None
+            is_invert = pattern_matcher(record.item)('!')
+            if is_invert and location.type != 'Song' and not world.shuffle_song_items:
+                ignore_pools = [2]
```
[the remark this answers is not in the document]
Comment (flagrama) [reply Plandomizer.py]: I mentioned some better method should be used such as an enum or something in the PR. Just not sure if that is the best method, or where to put it.

File: Plandomizer.py (hunk 1 above)
Comment (flagrama) [reply Plandomizer.py]: Seems to have worked fine. Done.

**Pull request merged**